qa/rpc-tests/getblocktemplate.py

```diff
@@ -36,40 +50,56 @@ def setup_network(self, split=False):
+            taddr0 = get_coinbase_address(node)
+            zaddr = node.z_getnewaddress('sapling')
+            recipients = [{"address": zaddr, "amount": Decimal('0.1')}]
+            myopid = node.z_sendmany(taddr0, recipients, 1, 0)
+            wait_and_assert_operationid_status(node, myopid)
+
+            # submit a tx with a transparent output
+            outputs = {node.getnewaddress():0.2}
+            node.sendmany('', outputs)
 
-        # Test 1: Default to coinbasetxn
-        tmpl = node.getblocktemplate()
-        assert('coinbasetxn' in tmpl)
-        assert('coinbasevalue' not in tmpl)
+        print("Getting block template")
+        gbt = node.getblocktemplate()
 
-        # Test 2: Get coinbasetxn if requested
-        tmpl = node.getblocktemplate({'capabilities': ['coinbasetxn']})
-        assert('coinbasetxn' in tmpl)
-        assert('coinbasevalue' not in tmpl)
+        prevhash = int(gbt['previousblockhash'], 16)
+        blockcommitmentshash = int(gbt['defaultroots']['blockcommitmentshash'], 16)
+        nTime = gbt['mintime']
+        nBits = int(gbt['bits'], 16)
 
-        # Test 3: coinbasevalue not supported if requested
-        tmpl = node.getblocktemplate({'capabilities': ['coinbasevalue']})
-        assert('coinbasetxn' in tmpl)
-        assert('coinbasevalue' not in tmpl)
+        f = BytesIO(hex_str_to_bytes(gbt['coinbasetxn']['data']))
+        coinbase = CTransaction()
+        coinbase.deserialize(f)
 
-        # Test 4: coinbasevalue not supported if both requested
-        tmpl = node.getblocktemplate({'capabilities': ['coinbasetxn', 'coinbasevalue']})
-        assert('coinbasetxn' in tmpl)
-        assert('coinbasevalue' not in tmpl)
+        print("Creating_block from template (simulating a miner)")
+        block = create_block(prevhash, coinbase, nTime, nBits, blockcommitmentshash)
 
-        # Test 5: General checks
-        tmpl = node.getblocktemplate()
-        assert_equal(16, len(tmpl['noncerange']))
-        # should be proposing height 2, since current tip is height 1
-        assert_equal(2, tmpl['height'])
+        # copy the non-coinbase transactions from the block template to the block
+        for gbt_tx in gbt['transactions']:
+            f = BytesIO(hex_str_to_bytes(gbt_tx['data']))
+            tx = CTransaction()
+            tx.deserialize(f)
+            block.vtx.append(tx)
+        block.hashMerkleRoot = int(gbt['defaultroots']['merkleroot'], 16)
+        assert_equal(block.hashMerkleRoot, block.calc_merkle_root(), "merkleroot")
+        assert_equal(len(block.vtx), len(gbt['transactions']) + 1, "number of transactions")
+        assert_equal(block.hashPrevBlock, int(gbt['previousblockhash'], 16), "prevhash")
+        block.solve()
+        block = block.serialize()
+        block = codecs.encode(block, 'hex_codec')
 
-        # Test 6: coinbasetxn checks
-        assert(tmpl['coinbasetxn']['required'])
+        print("Submitting block")
+        node.submitblock(block)
```

Comment:
Does this check that the block is valid?

Reply:
Good catch! I thought `submitblock` threw an exception in a failure case, but it does not, it return an error string or `None` if the call was successful. I've created PR #5437 to fix this (since this PR has already been merged). Thank you!